File: ProblemSolving/Algorithms/Search/Minimum_Loss.py
import itertools
#this may take time of  O(n*n) and an additional space of O(n)
def Naive(arr, n):
    #taking the base case
    if len(arr) > 1:
        min_diff = abs(arr[0] - arr[1])
    else:
        min_diff = 0
    loss = []   #having a list to take care of loss or
    # declare 2 variables for minloss and a loss
    for n1, n2 in itertools.combinations(arr, 2): 
        diff = n2-n1
        #print(diff)
        if diff < 0:
            loss.append(diff)
    return abs(max(loss))

def Algo(year):
    # sort the elements in the list first
    arr = sorted(year)
    # now find the minimum difference between the between the adjacent elements
    MinLoss = 10**10
    for i in range(1, len(arr)):
        if (arr[i] - arr[i-1] < MinLoss) and (year.index(arr[i]) < year.index(arr[i-1])):
            MinLoss = arr[i] - arr[i-1]

    return MinLoss



if __name__ == '__main__':

    n = int(input())
    arr = list(map(int, input().split()))

    #print(Naive(arr, n))
    print(Algo(arr))

File: ProblemSolving/Algorithms/Search/Minimum_Loss.cpp
#include <bits/stdc++.h>
using namespace std;

int MinimumLoss_Naive(vector<int>arr, int n)
{
    int diff;

    vector<int>loss;

    for(int i=0; i<n; i++){
        for(int j=i+1; j<n; j++){
            diff = arr[j] - arr[i];
            if(diff < 0)
                loss.push_back(diff);
        }
    }
    int maxdiff = INT_MIN;
    for(auto it=loss.begin(); it!=loss.end(); it++){
        //cout << *it << endl;
        if(maxdiff < *it)maxdiff = *it;
    }

    return abs(maxdiff);
}

int MinimumLoss_SearchAlgo(vector<int>arr, int n)
{
    /*
    The idea is to find the minimum loss(maximum negative integer)
    i.e from -5, -7, -2 -12 the minimum loss is -2(maximum negative integer)
    */
    vector<int>year;
    year = arr;         //copying the vector
    sort(arr.begin(), arr.end());
    int min_diff = INT_MIN;

    for(auto it = arr.begin()+1; it!=arr.end(); it++){
        if((*it - *(it-1) < min_diff) && (find(year.begin(), year.end(), *it) < find(year.begin(), year.end(), *(it-1)) ))
           min_diff = *it - *(it-1);
    }

    return min_diff;
}

int main()
{
    int n, temp;
    vector<int>arr;
    cin >> n;
     for(int i=0; i<n; i++){
        cin >> temp;
        arr.push_back(temp);
     }

     cout << MinimumLoss_Naive(arr, n);
     cout << MinimumLoss_SearchAlgo(arr, n);

    return 0;
}
